lib/lte_uart_dfu/lte_uart_dfu.c

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2022 INTELLINIUM <giuliano.franchetto@intellinium.com>
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <drivers/uart.h>
+#include <string.h>
+#include <lte_uart_dfu.h>
+#include <dfu/dfu_target_mcuboot.h>
+#include <dfu/dfu_target.h>
+#include <zephyr/logging/log.h>
+#include <stdnoreturn.h>
+#include <dfu_target_uart_encode.h>
+#include <dfu_target_uart_decode.h>
+#include <zephyr/sys/ring_buffer.h>
+
+LOG_MODULE_REGISTER(lte_uart_dfu, CONFIG_LTE_UART_DFU_LOG_LEVEL);
+
+K_THREAD_STACK_DEFINE(uart_handler_stack,
+		      CONFIG_LTE_UART_DFU_HANDLER_STACK_SIZE);
+static struct k_thread uart_handler_thread;
+static k_tid_t tid;
+static uint8_t in_buf[CONFIG_LTE_UART_DFU_IN_BUFFER_SIZE];
+
+struct lte_uart_dfu_ctx {
+	/* The UART device used for the DFU */
+	const struct device *uart_dev;
+
+	/* The ring buffer used to store received from the UART bus */
+	struct ring_buf rx_rb;
+
+	/* Semaphore used to wait data from UART bus */
+	struct k_sem lte_uart_sem;
+
+	/* Flag to check if the reception of application update has started */
+	bool download_started;
+	size_t file_size;
+	uint8_t image_num;
+	bool download_success;
+
+	/* Chuck */
+	const uint8_t *to_write;
+	size_t to_write_len;
+
+	/* The output buffer that will be sent to the host core */
+	uint8_t out_buffer[CONFIG_LTE_UART_DFU_OUT_BUFFER_SIZE];
+	size_t out_buf_cursor;
+} context;
+
+static int lte_uart_dfu_send_out_buffer(void)
+{
+	if (context.out_buf_cursor == 0) {
+		LOG_WRN("No data to send");
+		return 0;
+	}
+
+	LOG_DBG("Sending %zu bytes", context.out_buf_cursor);
+
+	for (int i = 0; i < context.out_buf_cursor; i++) {
+		uart_poll_out(context.uart_dev, context.out_buffer[i]);
+	}
+
+	context.out_buf_cursor = 0;
+
+	return 0;
+}
+
+static int lte_uart_dfu_send_error(int func, int error)
+{
+	int err;
+
+	const struct DFU_UART_target_RemoteResp fnc = {
+		.type_choice = func,
+		.error = error
+	};
+
+	err = cbor_encode_DFU_UART_target_RemoteResp(
+		context.out_buffer,
+		ARRAY_SIZE(context.out_buffer),
+		&fnc,
+		&context.out_buf_cursor);
+	if (err) {
+		LOG_ERR("Could not encode response, error %d", err);
+		return err;
+	}
+
+	return lte_uart_dfu_send_out_buffer();
+}
+
+static int lte_uart_dfu_send_offset(int error, size_t offset)
+{
+	int err;
+	int val = error == 0 ? (int) offset : error;
+
+	const struct DFU_UART_target_RemoteResp fnc = {
```

Comment:
Shouldn't there be separate responses for success, where offset is sent, and for error code?

Reply:
Could be, just tried to optimize the cddl by removing an optional ```?offset``` field that can be merge with the error field, and use the sign of the error field to know if this is an error (negative) or an offset (positive)